### PR TITLE
Adding preview builds

### PR DESCRIPTION
**Pull request by janosdebugs**
This PR adds preview builds for the oVirt website using GitHub Actions and GitHub pages. It works by building the oVirt site in a container, then copying the data, and finally pushing it to the `gh-pages` branch. This branch will need to be created empty.

When building, the GitHub action is configured to only run one job in parallel, which avoids problems with multiple jobs overwriting each other.

It also comments on the PR so contributors get their preview link. (This may possibly need a personal access token from a bot account.)

### CODE REVIEW

File: .github/workflows/preview.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        tags: ovirt-site:latest
+        cache-from: type=gha
+        cache-to: type=gha,mode=max
+    - name: Set up Jekyll cache
+      uses: actions/cache@v1
+      with:
+        path: vendor/bundle
+        key: ${{ runner.os }}-gems-${{ hashFiles('**/Gemfile.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-gems-
```
Comment (duck-rh) [inline .github/workflows/preview.yml:38]: IIUC this key does not exist and that would give the same result as not specifying `restore-keys`

File: .github/workflows/preview.yml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    - name: "Change config"
+      run: |
+        set -exuo pipefail
+        PR_ID=$(echo $GITHUB_REF | awk 'BEGIN { FS = "/" } ; { print $3 }')
+        PR_LINK=https://$(echo "${{github.repository}}" | sed -e 's#/#.github.io/#')/previews/${PR_ID}
```
Comment (duck-rh) [inline .github/workflows/preview.yml:43]: I would has created the path and base first and concatenated to make PR_LINK; I found this a bit confusing.

File: .github/workflows/remove-preview.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      cancel-in-progress: false
+    steps:
+    - name: Checkout gh-actions branch
+      uses: actions/checkout@v2
+      with:
+        ref: gh-pages
+    - name: Comment on PR
+      run: |
+        set -euo pipefail
+        PR_ID=$(echo $GITHUB_REF | awk 'BEGIN { FS = "/" } ; { print $3 }')
```
Comment (duck-rh) [inline .github/workflows/remove-preview.yml:21]: PR-ID is recalculated so many times and sometimes even inlined which does not help reading the code.
Maybe we could consider using one of the methods highlighted in: https://stackoverflow.com/questions/57819539/github-actions-how-to-share-a-calculated-value-between-job-steps
(as you did for targetdir)

**Review by duck-rh — CHANGES_REQUESTED** [review]
Looks pretty good, but added a few remarks.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    && echo "bundle install --quiet" >>/usr/local/bin/init.sh \
+    && echo 'exec bundle exec jekyll "$@"' >>/usr/local/bin/init.sh \
+    && chmod +x /usr/local/bin/init.sh
+
+WORKDIR /srv/site
+VOLUME /srv/site
+EXPOSE 4000
+
+ENTRYPOINT ["/usr/local/bin/init.sh"]
+CMD ["s","-t","-l","--host=0.0.0.0"]
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply Dockerfile:29]: Exposed  that port, want people to be able to do edits in their local env. For the build process the `CMD` is overridden anyway.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,20 @@
+FROM docker.io/library/ruby:2
+
+RUN DEBIAN_FRONTEND=noninteractive apt update \
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply Dockerfile:5]: Done, used setup.sh for this purpose.

File: .github/workflows/preview.yml (hunk 2 above)
Comment (janosdebugs) [reply .github/workflows/preview.yml]: Done.

File: .github/workflows/remove-preview.yml (hunk 3 above)
Comment (janosdebugs) [reply .github/workflows/remove-preview.yml]: Done.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,29 @@
+FROM docker.io/library/ruby:2
+
+# Copy and run the setup.
+COPY setup.sh /usr/local/bin/setup.sh
+RUN DEBIAN_FRONTEND=noninteractive apt update \
+    && apt install -y sudo \
+    && DEBIAN_FRONTEND=noninteractive SKIP_BUNDLE=1 /usr/local/bin/setup.sh \
+    && rm -rf /var/lib/apt/lists/*
```
Comment (duck-rh) [inline Dockerfile:8]: missing `apt clean` to remove downloaded debs in /var/cache/apt/archives/.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+# HTTP port.
+EXPOSE 4000
+
+# Live-reload port.
+EXPOSE 35729
+
+# All parameters passed to the container get routed through a script we create here.
+RUN echo "#!/bin/sh" >/usr/local/bin/init.sh \
+    && echo "set -e" >>/usr/local/bin/init.sh \
+    && echo "bundle config --local path 'vendor/bundle'" >>/usr/local/bin/init.sh \
```
Comment (duck-rh) [inline Dockerfile:24]: bundle config/install are already done in `setup.sh`
(and btw we do not need to repeat them in `run-server.sh` either; I often forget about this script and run the exec manually and that's really not needed)

**Review by duck-rh — CHANGES_REQUESTED** [review]
Looks really good.
The variables declaration makes things way easier to read.
I added minor comments and I think we're close to merging.

File: Dockerfile (hunk 7 above)
Comment (janosdebugs) [reply Dockerfile:24]: I made it conditional so that we can do it runtime rather than build time.

**Comment by duck-rh** [comment]
About dropping the scripts: I liked the idea to have the choice to use a container or not but keep the steps in common with both methods. I'll let @sandrobonazzola decide what is best, just my opinion.

I also spotted another problem. We do not build with ubuntu-lastest and I would not suggest using a fast-moving target for such project since we do not need bleeding edge features. Whatever target we choose in the future I think we should use the same base distro that is currently used to generate the production website. Even if the CI will catch any problem that would make development easier.

**Review by duck-rh — APPROVED** [review]
LGTM
It's a good idea to have brought all the checks there.

**Comment by janosdebugs** [comment]
@duck-rh I don't have rights to merge or retrigger the test run.

**Comment by duck-rh** [comment]
@janosdebugs I restarted the action but it fails due to the same problem.

**Comment by janosdebugs** [comment]
@duck-rh we'll need the machine user then.

**Comment by janosdebugs** [comment]
@sandrobonazzola help please?

**Review by sandrobonazzola — APPROVED** [review]
lgtn

**Comment by sandrobonazzola** [comment]
Checked with @janosdebugs ok to merge as the build is failing due to PR coming from a fork.

**Comment by duck-rh** [comment]
@sandrobonazzola OK

**Pull request merged**